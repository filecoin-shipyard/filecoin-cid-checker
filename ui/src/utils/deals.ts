import { FILECOIN_CID_CHECKER_API, PAGE_SIZE } from '../config/constants'
import { DealDetails, DealStatus, DealValue, DealValueNotAvailable } from './types'

export const truncateStringInTheMiddle = (
  str: string,
  strPositionStart: number,
  strPositionEnd: number,
) => {
  const minTruncatedLength = strPositionStart + strPositionEnd
  try {
    if (minTruncatedLength < str.length) {
      return `${str.substr(0, strPositionStart)}...${str.substr(
        str.length - strPositionEnd,
        str.length,
      )}`
    }
    return str
  } catch (e) {
    console.error(e)
  }
}

export const fetchDeals = async (search: string, page: number, query: string): Promise<DealValue[]> => {
  let url = `${FILECOIN_CID_CHECKER_API}deals?page=${page}&per_page=${PAGE_SIZE}`

  if (search) {
    url = `${FILECOIN_CID_CHECKER_API}deals/${search}?page=${page}&per_page=${PAGE_SIZE}`
  }

  url = url + query;
  const response = await fetch(url)
  const data = await response.json()

  const deals = data?.Deals ?? []

  const formattedDeals: DealValue[] = deals.map((deal: any) => {
    const { DealInfo } = deal

    const {
      PieceCID,
      Provider: MinerID,
      Client,
      PieceSize,
      VerifiedDeal,
      StartEpoch,
      EndEpoch,
      StoragePricePerEpoch,
      ProviderCollateral,
      ClientCollateral,
      Label,
    } = DealInfo?.Proposal
    const DealID = DealInfo?.DealID || DealValueNotAvailable
    const State = DealInfo.State.SectorStartEpoch > -1 ? DealStatus.Active : DealStatus.Unknown

    return {
      PieceCID: PieceCID['/'],
      DealID,
      MinerID,
      Client,
      PieceSize,
      VerifiedDeal,
      StartEpoch,
      EndEpoch,
      StoragePricePerEpoch,
      ProviderCollateral,
      ClientCollateral,
      State,
      Label,
    }
  })

  return formattedDeals
}

export const fetchDealDetails = async (dealId: string): Promise<DealDetails> => {
  const url = `${FILECOIN_CID_CHECKER_API}deals/details/${dealId}`
  const response = await fetch(url)
  const data = await response.json()

  return data
}
