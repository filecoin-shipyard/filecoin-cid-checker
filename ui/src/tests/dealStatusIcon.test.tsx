import React from 'react'
import { render } from '@testing-library/react'
import '@testing-library/jest-dom/extend-expect'

import { DealStatusIcon } from '../components/dealStatusIcon.component'
import { DealStatus } from '../utils/types'

describe('<DealStatusIcon /> Active', () => {
  it('renders the component', async () => {
    const { container } = render(<DealStatusIcon status={DealStatus.Active} />)

    const headerClass = DealStatusIcon({ status: DealStatus.Active }).type.styledComponentId
    const MyHeaderRoots = document.getElementsByClassName(headerClass)
    const style = window.getComputedStyle(MyHeaderRoots[0])

    expect(style.backgroundColor).toBe('rgb(61, 177, 74)')
    expect(container).toMatchSnapshot()
  })
})

describe('<DealStatusIcon /> Fail', () => {
  it('renders the component', async () => {
    const { container } = render(<DealStatusIcon status={DealStatus.Fault} />)

    const headerClass = DealStatusIcon({ status: DealStatus.Fault }).type.styledComponentId
    const MyHeaderRoots = document.getElementsByClassName(headerClass)
    const style = window.getComputedStyle(MyHeaderRoots[0])

    expect(style.backgroundColor).toBe('rgb(214, 196, 30)')
    expect(container).toMatchSnapshot()
  })
})

describe('<DealStatusIcon /> Active', () => {
  it('renders the component', async () => {
    const { container } = render(<DealStatusIcon status={DealStatus.Recovery} />)

    const headerClass = DealStatusIcon({ status: DealStatus.Fault }).type.styledComponentId
    const MyHeaderRoots = document.getElementsByClassName(headerClass)
    const style = window.getComputedStyle(MyHeaderRoots[0])

    expect(style.backgroundColor).toBe('rgb(136, 148, 168)')

    expect(container).toMatchSnapshot()
  })
})

describe('<DealStatusIcon /> Active', () => {
  it('renders the component', async () => {
    const { container } = render(<DealStatusIcon status={DealStatus.Unknown} />)

    const headerClass = DealStatusIcon({ status: DealStatus.Fault }).type.styledComponentId
    const MyHeaderRoots = document.getElementsByClassName(headerClass)
    const style = window.getComputedStyle(MyHeaderRoots[0])

    expect(style.backgroundColor).toBe('rgb(240, 41, 26)')

    expect(container).toMatchSnapshot()
  })
})
