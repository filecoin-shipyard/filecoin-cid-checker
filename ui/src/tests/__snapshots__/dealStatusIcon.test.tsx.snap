// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<DealStatusIcon /> Active renders the component 1`] = `
<div>
  <div
    class="sc-AxjAm gNGPFX"
  />
</div>
`;

exports[`<DealStatusIcon /> Active renders the component 2`] = `
<div>
  <div
    class="sc-AxjAm wieiO"
  />
</div>
`;

exports[`<DealStatusIcon /> Active renders the component 3`] = `
<div>
  <div
    class="sc-AxjAm eeLvrn"
  />
</div>
`;

exports[`<DealStatusIcon /> Fail renders the component 1`] = `
<div>
  <div
    class="sc-AxjAm coWhuT"
  />
</div>
`;
